README
Net-APNS-Persistent version 0.03
================================

Net::APNS::Persistent - Send Apple APNS notifications over a persistent connection
Net::APNS::Feedback - Retrieve data from Apple's APNS feedback service

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Class::Accessor
  Net::SSLeay
  Socket
  Encode
  JSON::XS
  Test::More
  Test::Exception
  FindBin

USAGE

Please refer to the POD documentation in the perl modules themselves.

SEE ALSO

Homepage for this module:
  http://github.com/aufflick/p5-net-apns-persistent

Apple APNS documentation:
  http://developer.apple.com/IPhone/library/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Introduction/Introduction.html


COPYRIGHT AND LICENCE

Copyright (C) 2009 by Mark Aufflick

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.9 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension Net::APNS::Persistent and Net::APNS::Feedback.

0.03  Thur Sep 30 2010
    - Raises exception on send (instead of infinite loop) if packet is oversize due
      to custom payload content

0.02  Mon Dec 07 2009
    - Fixes rt #52517
      Some versions of perl 5.8 mandate a* after / in pack()

0.01  Sun Nov 29 15:17:29 2009
	- First version of both modules



lib/Net/APNS/Feedback.pm
package Net::APNS::Feedback;

use 5.008;
use strict;
use warnings;

our $VERSION = '0.03';

use base 'Net::APNS::Persistent::Base';

use JSON::XS;

# ensure we're in byte-oriented mode
use bytes;

my %defaults = (
    host_production => 'feedback.push.apple.com',
    host_sandbox    => 'feedback.sandbox.push.apple.com',
    port            => 2196,
   );

=head1 NAME

Net::APNS::Feedback - Retrieve data from Apple's APNS feedback service

=head1 SYNOPSIS

  use Net::APNS::Feedback;
  
  my $apns = Net::APNS::Feedback->new({
    sandbox => 1,
    cert    => 'cert.pem',
    key     => 'key.pem',
    passwd  => 'key password',
  });
  
  my @feedback = $apns->retrieve_feedback;

=head1 DESCRIPTION

Apple's APNS system provides a feedback service to let you know the
device rejected notifications because they are no longer wanted
(usually meaning the app has been removed).

L<See the|SEE ALSO> Apple Push Notification Service Programming Guide.

=head1 METHODS

=head2 new

Args:

=over

=item sandbox

set to true if you want to use the sandbox host. defaults to 0. ignored if you set the host manually

=item cert

path to your certificate

=item cert_type

defaults to PEM - see L<Net::SSLeay>.

=item key

path you your private key

=item key_type

defaults to PEM - see L<Net::SSLeay>.

=item passwd

password for your private key, if required.

=item host

defaults to feedback.push.apple.com or feedback.sandbox.push.apple.com depending
on the setting of sandbox. can be set manually.

=item port

defaults to 2196

=back

NB: all these args are available as accessors, but you need to set them before the connection
is first used.

=cut

sub new {
    my ($class, $init_vals) = @_;

    $init_vals ||= {};

    my $self = $class->SUPER::new({

        %defaults,
        
        %{$init_vals}
       });

    return $self;
}

=head2 retrieve_feedback

Takes no arguments and returns an arrayref (possibly) containing hashrefs. eg:

  [
    {
      'time_t' => 1259577923,
      'token' => '04ef31c86205...624f390ea878416'
    },
    {
      'time_t' => 1259577926,
      'token' => '04ef31c86205...624f390ea878416'
    },
  ]

C<time_t> is the epoc time of when the notification was rejected. C<token> is
a hex encoded device token for you to reconcile with your data.

As you can see from this example, you can recieve more than one notifications
about the same token if you have had more than one message rejected since you last
checked the feedback service.

Note that once you have drained all the feedback, you will not be delivered the
same set again.

L<See the|SEE ALSO> Apple Push Notification Service Programming Guide.

=cut

sub retrieve_feedback {
    my $self = shift;

    my $data = $self->_read;

    my @res;

    while ($data) {
        my ($time_t, $token_bin);
        ($time_t, $token_bin, $data) = unpack( 'N n/a a*', $data);

        push @res, {
            time_t => $time_t,
            token => unpack( 'H*', $token_bin ),
        };
    }

    return \@res;
}

=head2 disconnect

Disconnect the ssl connection and socket, and free the ssl structures. This usually
isn't necessary as this will happen implicitly when the object is destroyed.

=head1 SEE ALSO

=over 4

=item Presentation on this module by Author

L<http://mark.aufflick.com/talks/apns>

=item Apple Push Notification Service Programming Guide

L<http://developer.apple.com/IPhone/library/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Introduction/Introduction.html>

=item L<Net::APNS::Persistent>

=item GIT Source Repository for this module

L<http://github.com/aufflick/p5-net-apns-persistent>

=back

=head1 AUTHOR

Mark Aufflick, E<lt>mark@aufflick.comE<gt>, L<http://mark.aufflick.com/>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Mark Aufflick

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.9 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/Net/APNS/Persistent/Base.pm
package Net::APNS::Persistent::Base;

use 5.008;
use strict;
use warnings;

our $VERSION = '0.03';

use base 'Class::Accessor';

use Net::SSLeay qw(die_now die_if_ssl_error);
use Socket;

__PACKAGE__->mk_accessors(qw(
                                cert
                                key
                                key_type
                                cert_type
                                passwd
                                _host
                                host_production
                                host_sandbox
                                port
                                sandbox
                                __connection
                           ));

my %defaults = (
    key_type => &Net::SSLeay::FILETYPE_PEM,
    cert_type => &Net::SSLeay::FILETYPE_PEM,
   );

=head1 NAME

Net::APNS::Persistent::Base - Base class for Net::APNS::Persistent and Net::APNS::Feedback

=head1 SYNOPSIS

See L<Net::APNS::Persistent> and L<Net::APNS::Feedback>.

=head1 DESCRIPTION

Base methods for L<Net::APNS::Persistent> and L<Net::APNS::Feedback>.

=cut

sub new {
    my ($class, $init_vals) = @_;

    $init_vals ||= {};

    my $self = $class->SUPER::new({

        %defaults,
        
        %{$init_vals}
       });

    # Use TLSv1
    $Net::SSLeay::ssl_version = 10;

    Net::SSLeay::load_error_strings();
    Net::SSLeay::SSLeay_add_ssl_algorithms();
    Net::SSLeay::randomize();
    
    return $self;
}

sub host {
    my $self = shift;

    if (@_ || $self->_host) {
        return $self->_host(@_);
    }

    if ($self->sandbox) {
        return $self->host_sandbox;
    } else {
        return $self->host_production;
    }
}

sub _server_sockadder_in {
    my $self = shift;

    my $ip = inet_aton( $self->host )
      or die "Unable to get address for " . $self->host;

    return sockaddr_in( $self->port, $ip );
}

sub _connection {
    my $self = shift;
    
    my $connection = $self->__connection;

    my ($socket, $ctx, $ssl);
    ($socket, $ctx, $ssl) = @{$connection}
      if $connection;
    
    # TODO: attempt reconnect if the socket is bad
    # at the moment, will just die which is ok since
    # we only promise to raise an exception in the doccs.
    if (!$socket || !$ctx || !$ssl) {
        
        # free any existing resources
        $self->disconnect;
        
        socket( $socket, PF_INET, SOCK_STREAM, getprotobyname('tcp'))
          or die "error creating socket: $!";

        connect( $socket, $self->_server_sockadder_in )
          or die "error connecting socket: $!";

        $ctx = Net::SSLeay::CTX_new()
          or die_now( "failed to create SSL_CTX: $!");

        Net::SSLeay::CTX_set_options( $ctx, &Net::SSLeay::OP_ALL );
        die_if_ssl_error("error while setting ctx options: $!");

        Net::SSLeay::CTX_set_default_passwd_cb( $ctx, sub { $self->passwd } );
        Net::SSLeay::CTX_use_RSAPrivateKey_file( $ctx, $self->key, $self->key_type );
        die_if_ssl_error("error while setting private key: $!");

        Net::SSLeay::CTX_use_certificate_file( $ctx, $self->cert, $self->cert_type );
        die_if_ssl_error("error while setting certificate: $!");

        $ssl = Net::SSLeay::new( $ctx );

        Net::SSLeay::set_fd( $ssl, fileno($socket) );
        Net::SSLeay::connect( $ssl )
            or die_now( "failed ssl connect: $!");

        $connection = [$socket, $ctx, $ssl];
        
        $self->__connection($connection);
    }

    return $connection;
}

sub disconnect {
    my $self = shift;
    
    # go straight to the accessor, we don't
    # want to create a connection simply to
    # disconnect it!
    my $connection = $self->__connection;

    if (!$connection) {
        return 1;
    }

    my ($socket, $ctx, $ssl) = @{$connection};

    CORE::shutdown( $socket, 1 )
        if $socket;
    
    Net::SSLeay::free( $ssl )
        if $ssl;
    
    Net::SSLeay::CTX_free( $ctx )
        if $ctx;
    
    close($socket)
      if $socket;

    $self->__connection(undef);

    return 1;
}

sub _send {
    my ($self, $data) = @_;
    
    my ($socket, $ctx, $ssl) = @{$self->_connection};

    Net::SSLeay::ssl_write_all( $ssl, $data );
    die_if_ssl_error("error writing to ssl connection: $!");

    return 1;
}

sub _read {
    my $self = shift;

    my ($socket, $ctx, $ssl) = @{$self->_connection};

    my $data = Net::SSLeay::ssl_read_all( $ssl );
    die_if_ssl_error("error reading from ssl connection: $!");

    return $data;
}

sub DESTROY {
    my $self = shift;

    $self->disconnect;
}



=head1 SEE ALSO

=over 4

=item L<Net::APNS::Persistent>

=item L<Net::APNS::Feedback>.

=item L<http://github.com/aufflick/p5-net-apns-persistent|GIT Source Repository for this module>

=back

=head1 AUTHOR

Mark Aufflick, E<lt>aufflick@localE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Mark Aufflick

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.9 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;


lib/Net/APNS/Persistent.pm
package Net::APNS::Persistent;

# perl 5.8 required for utf8 safe substr
use 5.008;
use strict;
use warnings;

our $VERSION = '0.03';

use base 'Net::APNS::Persistent::Base';

use Encode qw(decode encode encode_utf8);

# NB: Using JSON::XS as not all JSON modules allowed
# by JSON::Any do unicode correctly
use JSON::XS;

__PACKAGE__->mk_accessors(qw(
                                devicetoken
                                _queue
                                max_payload_size
                                command
                                _json
                           ));

my %defaults = (
    host_production  => 'gateway.push.apple.com',
    host_sandbox     => 'gateway.sandbox.push.apple.com',
    max_payload_size => 256,
    port             => 2195,
    command          => 0
   );

=head1 NAME

Net::APNS::Persistent - Send Apple APNS notifications over a persistent connection

=head1 SYNOPSIS

  use Net::APNS::Persistent;

  my $devicetoken_hex = '04ef...a878416';
  
  my $apns = Net::APNS::Persistent->new({
    sandbox => 1,
    cert    => 'cert.pem',
    key     => 'key.pem',
    passwd  => 'key password',
  });

  $apns->queue_notification(
    $devicetoken_hex,
    {
      aps => {
          alert => 'sweet!',
          sound => 'default',
          badge => 1,
      },
    });
  
  $apns->send_queue;

  $apns->disconnect;

You can queue more than one notification in one transmission
by calling L<queue_notification> multiple times. If you want to
pass in utf8 text in the alert (either as a string or alert-E<gt>body),
you need to be careful with the encoding. See the test files for an
example of reading utf8 from a text file. You should also be able
to pass utf8 through from eg. a database in a similar way.

You can also use the connection many times (ie. queue then send, queue then send,
ad nauseum). The call to disconnect is not strictly necessary since the object
will disconnect as soon as it falls out of scope.

You can place your own custom data outside the C<aps> hash. L<See the|SEE ALSO>
Apple Push Notification Service Programming Guide for more info.

All methods are fatal on error. Eg. if the ssl connection returns an error,
the code will die. You can either then just restart your script or you can use
C<eval> to catch the exception.

=head1 DESCRIPTION

Class to create a persistent connection to Apple's APNS servers

=head1 METHODS

=head2 new

Args:

=over

=item sandbox

set to true if you want to use the sandbox host. defaults to 0. ignored if you set the host manually

=item cert

path to your certificate

=item cert_type

defaults to PEM - see L<Net::SSLeay>.

=item key

path you your private key

=item key_type

defaults to PEM - see L<Net::SSLeay>.

=item passwd

password for your private key, if required.

=item host

defaults to gateway.push.apple.com or gateway.sandbox.push.apple.com depending
on the setting of sandbox. can be set manually.

=item port

defaults to 2195

=item command

defaults to 0

=back

NB: all these args are available as accessors, but you need to set them before the connection
is first used.

=cut

sub new {
    my ($class, $init_vals) = @_;

    $init_vals ||= {};

    my $self = $class->SUPER::new({

        %defaults,
        
        %{$init_vals}
       });

    $self->_queue([]);

    $self->_json(JSON::XS->new());
    $self->_json->utf8(1);

    return $self;
}

sub _apply_to_alert_body {
    my ($payload, $func) = @_;

    return
      if ! exists $payload->{aps}{alert};
    
    # can be in alert->body, or a plain string in alert
    if (ref $payload->{aps}{alert} eq 'HASH') {
        $payload->{aps}{alert}{body} = $func->($payload->{aps}{alert}{body});
    } else {
        $payload->{aps}{alert} = $func->($payload->{aps}{alert});
    }
}

sub _alert_body {
    my ($payload) = @_;

    return $payload->{aps}{alert}{body}
        if ref $payload->{aps}{alert} eq 'HASH';

    return $payload->{aps}{alert};
}

sub _pack_payload_for_devicetoken {
    my ($self, $devicetoken, $payload) = @_;

    if (ref($payload) ne 'HASH' || ref($payload->{aps}) ne 'HASH') {
        die "Invalid payload: " . Dumper($payload);
    }

    # force badge to be integer
    $payload->{aps}{badge} += 0
      if exists $payload->{aps}{badge};

    # convert message to unicode, after ensuring it was utf8 in the first place
    _apply_to_alert_body($payload, sub {
                             my $str = shift; # decode won't work on string literals
                             encode('unicode', decode('utf8', $str, Encode::FB_CROAK));
                         });

    my $json = $self->_json->encode($payload);

    # enforce max_payload_size
    my $max_payload_size = $self->max_payload_size;
    if ( bytes::length($json) > $max_payload_size ) {
        
        # not sure why this is necessary. Must be something
        # about the difference in density b/n utf8 and unicode?
        # This isn't very efficient,
        # but users shouldn't be passing in huge strings, surely...
        
        while (bytes::length($json) > $max_payload_size) {

            if (!_alert_body($payload)) {
                die "Cannot bring payload under $max_payload_size bytes by truncating alert string";
            }
            
            _apply_to_alert_body($payload, sub {
                                     substr($_[0], 0, -1);
                                 });

            $json = JSON::XS::encode_json($payload);
        }
    }

    return pack(
        'c n/a* n/a*',
        $self->command,
        pack( 'H*', $devicetoken ),
        $json
       );
}

=head2 queue_notification

takes two arguments - a device token (as a string representation of hex), and
a hashref with the payload. eg:

  my $devicetoken_hex = '04ef...a878416';

  $apns->queue_notification(
    $devicetoken_hex,
    {
      aps => {
          alert => 'sweet!',
          sound => 'default',
          badge => 1,
      },
    });

  $apns->queue_notification(
    $devicetoken_hex,
    {
      aps => {
          alert => {
              body => 'foo',
              'action-loc-key' => undef,
          },
          sound => 'default',
          badge => 1,
      },
      foo => 'bar',
    });

The second example shows the complex alert format and also custom application
data outside the aps hash.

This method will ensure that the payload is at most 256 bytes by trimming the
alert body. The trimming function is utf8-safe, but not very efficient (so
don't ask it to trim War and Peace).

=cut

sub queue_notification {
    my ($self, $devicetoken, $payload) = @_;

    push @{$self->_queue}, [$devicetoken, $payload];

    return 1;
}

=head2 send_queue

This will actually send the data to the ssl connection.

=cut

sub send_queue {
    my $self = shift;

    my $data = '';

    for my $queue_entry (@{$self->_queue}) {
        my ($devicetoken, $payload) = @{$queue_entry};

        $data .= $self->_pack_payload_for_devicetoken($devicetoken, $payload);
    }

    $self->_send($data)
      if $data;

    $self->_queue([]);

    return 1;
}

=head2 disconnect

Disconnect the ssl connection and socket, and free the ssl structures. This usually
isn't necessary as this will happen implicitly when the object is destroyed.

=head1 SEE ALSO

=over 4

=item Presentation on this module by Author

L<http://mark.aufflick.com/talks/apns>

=item Apple Push Notification Service Programming Guide

L<http://developer.apple.com/IPhone/library/documentation/NetworkingInternet/Conceptual/RemoteNotificationsPG/Introduction/Introduction.html>

=item L<Net::APNS::Feedback>

=item GIT Source Repository for this module

L<http://github.com/aufflick/p5-net-apns-persistent>

=back

=head1 AUTHOR

Mark Aufflick, E<lt>mark@aufflick.comE<gt>, L<http://mark.aufflick.com/>

=head1 CREDITS

Some inspiration came from haoyayoi's L<Net::APNS>.

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2009 by Mark Aufflick

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.8.9 or,
at your option, any later version of Perl 5 you may have available.

=cut

1;
